workflow/controller/operator.go

```diff
@@ -3938,3 +3968,22 @@ func setWfPodNamesAnnotation(wf *wfv1.Workflow) {
 
 	wf.Annotations[common.AnnotationKeyPodNameVersion] = podNameVersion.String()
 }
+
+func getChildNodeIdsAndLastRetriedNode(node *wfv1.NodeStatus, nodes wfv1.Nodes) ([]string, *wfv1.NodeStatus) {
+	if len(node.Children) == 0 {
+		return []string{}, nil
+	}
+
+	childNodeIds := getChildNodeIdsRetried(node, nodes)
+
+	// We could not find child nodes, we are trying to retry hooked node itself.
+	if len(childNodeIds) == 0 {
+		childNodeIds = node.Children
+	}
+
+	lastChildNode, err := nodes.Get(childNodeIds[len(childNodeIds)-1])
+	if err != nil {
+		panic(fmt.Sprintf("could not find nodeId %s in Children of node %+v", childNodeIds[len(childNodeIds)-1], node))
+	}
+	return childNodeIds, lastChildNode
+}
```

Comment:
please add a comment to the top of this function

Reply:
I really hate to mention this - but I'm debating if in addition to the "Hooked" boolean you added, we should have a "Retried" boolean? And maybe the logic should be based on that instead? Would it make this method more readable? (sorry) What do you think?

Reply:
I guess it's mainly that I find line 3981 unclear. If you think there's a good way to make it more clear through comments, then we don't need to add the "Retried" boolean. thank you!

Reply:
I'm inclining to implement `NodeStatus.Retried` (bool) for future expansion.
What do you think on keeping the `Hooked` field? We could delete `Hooked` if we add `Retried` field to resolve this issue.
Or you would like to keep `Hooked` field for clarification?

Just FYI:

I've added that line(3981) for retrying `Hooked` node where all `Retry` Node children are marked as `Hooked`.
I can write a comment on this, but `Retried` NodeStatus should be more clear and extensive.

![image](https://github.com/argoproj/argo-workflows/assets/83329336/7c37d597-d320-4eca-bd23-5f2545f91d57)

<details>
<summary>sample yaml</summary>

```yaml
spec:
  templates:
    - name: argosay
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: argoproj/argosay:v2
        command:
          - /bin/sh
          - '-c'
        args:
          - /bin/sleep 2; exit 100; /argosay
        resources: {}
      retryStrategy:
        limit: 2
    - name: argosay2
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: argoproj/argosay:v2
        command:
          - /bin/sh
          - '-c'
        args:
          - sleep 2; exit 0
        resources: {}
  entrypoint: argosay2
  arguments: {}
  hooks:
    failed:
      template: argosay
      arguments: {}
      expression: workflow.status == "Failed"
    running:
      template: argosay
      arguments: {}
      expression: workflow.status == "Running"
    succeed:
      template: argosay
      arguments: {}
      expression: workflow.status == "Succeeded"
```

</details>

Reply:
I should fix some unit tests, but added `NodeStatus.NodeFlag` to clarify node is hooked, retried, etc.
[`b0d7e6d` (#11839)](https://github.com/argoproj/argo-workflows/pull/11839/commits/b0d7e6d26d544fb3575633ad83f7d5c823d10ed7#diff-94d0f1284fafd0a1cb05dc2309b4be043b6fb01d66626ab1867d40c84b573ec2https)

I hope you like this idea :).

Reply:
great, thanks so much for reworking this a few times! I think `NodeFlag` is good. :)

Reply:
I've fixed some unit tests and e2e tests.
And added a description in PR for a complicated case we have discussed.
Kindly review them :).